src/osd/PrimaryLogPG.cc

```diff
@@ -3717,6 +3717,39 @@ int PrimaryLogPG::do_xattr_cmp_str(int op, string& v1s, bufferlist& xattr)
   }
 }
 
+int PrimaryLogPG::do_extent_cmp(OpContext *ctx, OSDOp& osd_op)
+{
+  ceph_osd_op& op = osd_op.op;
+  vector<OSDOp> read_ops(1);
+  OSDOp& read_op = read_ops[0];
+  int result = 0;
+
+  read_op.op.op = CEPH_OSD_OP_SYNC_READ;
+  read_op.op.extent.offset = op.extent.offset;
+  read_op.op.extent.length = op.extent.length;
+  read_op.op.extent.truncate_seq = op.extent.truncate_seq;
+  read_op.op.extent.truncate_size = op.extent.truncate_size;
+
+  result = do_osd_ops(ctx, read_ops);
+  if (result < 0) {
+    derr << "do_extent_cmp do_osd_ops failed " << result << dendl;
+    return result;
+  }
+
+  if (read_op.outdata.length() != osd_op.indata.length())
+    return -EINVAL;
+
+  for (uint64_t p = 0; p < osd_op.indata.length(); p++) {
+    if (read_op.outdata[p] != osd_op.indata[p]) {
+      ::encode(p, osd_op.outdata);
+      osd_op.outdata.claim_append(read_op.outdata);
```

Comment:
Got these review comments from Jason and Sage at Vault:

- The client may retry this request, because for example it lost the connection and did not get the response, and we could end up getting the correct return/error value (-ELISEQ), but the returned data buffer may contain different data than the first try.

- SCSI COMPARE_AND_WRITE is the only known user of this op and it only needs the miscompare offset. We can just return this as the return value (0 is success, ret > 0 is the offset in the buffer the miscompare occurred, and ret < 0 is a -Exyz error).